dwisc.py

```diff
@@ -100,54 +100,71 @@ def main(args):
-
-            submitted_problems = []
-            for i in range(args.calls_per_round):
-                num_reads = min(args.call_num_reads, num_reads_remaining)
-                params['num_reads'] = num_reads
-
-                print_err('    submit {} of {} remaining'.format(num_reads, num_reads_remaining))
-
-                submitted_problems.append({
-                    'problem': solver.sample_ising(h, J, **params),
-                    'start_time': datetime.datetime.utcnow(),
-                    'params': {k:v for k,v in params.items()}
-                    })
-                num_reads_remaining -= num_reads
-
-                if num_reads_remaining <= 0:
-                    break
-
-            #answers = solve_ising(solver, h, J, **params)
-            print_err('    waiting...')
-
-            for i, submitted_problem in enumerate(submitted_problems):
-                problem = submitted_problem['problem']
-                problem.wait()
-                print_err('    collect {} of {} calls'.format(i+1, len(submitted_problems)))
-                answers = problem.result()
-
-                solutions = answers_to_solutions(
-                    answers,
-                    data['variable_ids'],
-                    submitted_problem['start_time'],
-                    datetime.datetime.utcnow(),
-                    submitted_problem['params'],
-                    solution_metadata
-                )
-                if solutions_all != None:
-                    combis.combine_solution_data(solutions_all, solutions)
+            try:
+                print_err('')
+                print_err('  collection round {} of {} (sample_ising calls per round {})'.format(iteration, rounds, args.calls_per_round))
+
+                submitted_problems = []
+                num_reads_remaining_prev = num_reads_remaining
+                for i in range(args.calls_per_round):
+                    num_reads = min(args.call_num_reads, num_reads_remaining)
+                    params['num_reads'] = num_reads
+
+                    print_err('    submit {} of {} remaining'.format(num_reads, num_reads_remaining))
+
+                    submitted_problems.append({
+                        'problem': solver.sample_ising(h, J, **params),
+                        'start_time': datetime.datetime.utcnow(),
+                        'params': {k:v for k,v in params.items()}
+                        })
+                    num_reads_remaining -= num_reads
+
+                    if num_reads_remaining <= 0:
+                        break
+
+                #answers = solve_ising(solver, h, J, **params)
+                print_err('    waiting...')
+
+                for i, submitted_problem in enumerate(submitted_problems):
+                    problem = submitted_problem['problem']
+                    if problem.wait(timeout = 300) is False:
```

Comment:
I have observed this "hang indefinitely" problem as well.  Setting a suitable timeout is a little tricky as it will depends on how heavily loaded the machine is, so this should be able to be set via a command line argument.  300 seconds seems like a suitable default value.